File: codeforces/edu135/D.go
// Read lines from stdin using bufio
package main

import (
	"bufio"
	"fmt"
	"os"
	"math"
	// "strings"
)

func dfs(s string, m map[string]int) int {
	// if the length of s is 0
	if len(s) == 0 {
		return 0
	}
	if _, ok := m[s]; ok {
		return m[s]
	}
	results := make([]int, 0)
	// Create an array that holds pairs of two strings
	pairs := make([][]string, 0)
	// Append the first character of s and s without the first character to the array
	pairs = append(pairs, []string{s[:1], s[1:]})
	pairs = append(pairs, []string{s[len(s)-1:], s[:len(s)-1]})
	// Iterate through the array
	for _, pair := range pairs {
		alice := pair[0]
		remaining := pair[1]
		var a int
		var b int
		a = dfs(remaining[1:], m)
		b = dfs(remaining[:len(remaining)-1], m)
		// set the variable res to the maximum of a and b
		res := math.Max(float64(a), float64(b))
		if res != 0 {
			results = append(results, int(res))
		} else {
			if a >= 0{
				if alice[0] > remaining[0]{
					results = append(results, -1)
				} else if alice[0] < remaining[0] {
					results = append(results, 1)
				} else {
					results = append(results, 0)
				}
			} else {
				if alice[0] > remaining[len(remaining)-1]{
					results = append(results, -1)
				} else if alice[0] < remaining[len(remaining)-1] {
					results = append(results, 1)
				} else {
					results = append(results, 0)
				}
			}
		}
	}
	// return the minimum of the array results
	var answer = float64(results[0])
	for _, result := range results {
		answer = math.Min(answer, float64(result))
	}
	m[s] = int(answer)
	return int(answer)
}

func main() {
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()
	var n int
	// read int from input to a variable
	fmt.Fscan(in, &n)
	for i := 0; i < n; i++ {
		// read one line from input to a variable s
		var s string
		fmt.Fscan(in, &s)
		// Create a hashmap from string to int
		var m = make(map[string]int)
		res := dfs(s, m)
		if res == 0 {
			fmt.Println("Draw")
		} else {
			fmt.Println("Alice")
		}
	}
}

File: codeforces/edu135/D.py
import sys
from functools import lru_cache
lines = list(map(str.strip, sys.stdin.readlines()))
from collections import deque

def comp(a, b):
    if a < b: return -1
    if a > b: return 1
    return 0

for line in lines[1:]:
    n = len(line)
    s = line
    dp = [[0 for _ in range(n+5)] for _ in range(n+5)]
    for length in range(2, n+1, 2):
        for l in range(n-length+1):
            r = l + length
            dp[l][r] = 1
            res = -1
            if dp[l+1][r-1] != 0:
                res = max(res, dp[l+1][r-1])
            else:
                res = max(res, comp(s[l], s[r-1]))
            if dp[l+2][r] != 0:
                res = max(res, dp[l+2][r])
            else:
                res = max(res, comp(s[l], s[l+1]))
            dp[l][r] = min(dp[l][r], res)
            res = -1
            if dp[l+1][r-1] != 0:
                res = max(res, dp[l+1][r-1])
            else:
                res = max(res, comp(s[r-1], s[l]))
            if dp[l][r-2] != 0:
                res = max(res, dp[l][r-2])
            else:
                res = max(res, comp(s[r-1], s[r-2]))
            dp[l][r] = min(dp[l][r], res)
    print("Alice" if dp[0][n] == -1 else "Draw")